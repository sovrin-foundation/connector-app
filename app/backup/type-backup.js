// @flow
import type { ReactNavigation } from '../common/type-common'
import type { CustomError } from '../common/type-common'
import type { IsValid } from '../components/input-control/type-input-control'

export type ReactNavigationBackup = {
  navigation: {
    navigate: (route: string, params?: any) => void,
    goBack: (route?: ?string) => void,
    state: {
      params: {
        recoveryPassphrase?: string,
        initialRoute: string,
      },
    },
  },
}

export type GenerateRecoveryPhraseProps = {
  generateRecoveryPhrase: () => void,
  recoveryPassphrase: Passphrase,
  recoveryStatus: $Keys<typeof BACKUP_STORE_STATUS>,
} & ReactNavigationBackup

export type GenerateRecoveryPhraseState = {
  chatBubbleDimensions: ChatBubbleDimensions,
}

export type ChatBubbleDimensions = {
  width: number,
  height: number,
  x: number,
  y: number,
}

export type PassphraseTextProps = {
  recoveryPassphrase: Passphrase,
  chatBubbleDimensions: ChatBubbleDimensions,
}

export type PassphraseErrorProps = {
  chatBubbleDimensions: ChatBubbleDimensions,
}

export type VerifyRecoveryPhraseProps = {
  recoveryPassphrase: Passphrase,
} & ReactNavigationBackup

export type VerifyRecoveryPhraseState = {
  error: boolean,
}

export type ExportBackupFileProps = {
  backupPath: string,
  backupStatus: string,
  exportBackup: () => void,
} & ReactNavigationBackup

export type BackupErrorProps = {
  updateStatusBarTheme: string => void,
  generateBackupFile: () => void,
} & ReactNavigationBackup

export type ExportBackupFileState = {
  submitButtonText: string,
}

export type BackupCompleteProps = {} & ReactNavigationBackup

export type BackupCompleteState = {
  submitButtonText: string,
  recoveryPassphrase: string,
}

export const PREPARE_BACK_IDLE = 'PREPARE_BACKUP_IDLE'
export const PREPARE_BACKUP_LOADING = 'PREPARE_BACKUP_LOADING'
export const PREPARE_BACKUP_SUCCESS = 'PREPARE_BACKUP_SUCCESS'
export const PREPARE_BACKUP_FAILURE = 'PREPARE_BACKUP_FAILURE'

export const BACKUP_STORE_STATUS = {
  IDLE: 'IDLE',
  GENERATE_PHRASE_LOADING: 'GENERATE_PHRASE_LOADING',
  GENERATE_PHRASE_SUCCESS: 'GENERATE_PHRASE_SUCCESS',
  GENERATE_PHRASE_FAILURE: 'GENERATE_PHRASE_FAILURE',
  GENERATE_BACKUP_FILE_LOADING: 'GENERATE_BACKUP_FILE_LOADING',
  GENERATE_BACKUP_FILE_SUCCESS: 'GENERATE_BACKUP_FILE_SUCCESS',
  GENERATE_BACKUP_FILE_FAILURE: 'GENERATE_BACKUP_FILE_FAILURE',
  PREPARE_BACKUP_LOADING,
  PREPARE_BACKUP_SUCCESS,
  PREPARE_BACKUP_FAILURE,
  PREPARE_BACK_IDLE,
  EXPORT_BACKUP_LOADING: 'EXPORT_BACKUP_LOADING',
  EXPORT_BACKUP_SUCCESS: 'EXPORT_BACKUP_SUCCESS',
  EXPORT_BACKUP_FAILURE: 'EXPORT_BACKUP_FAILURE',
  EXPORT_BACKUP_NO_SHARE: 'EXPORT_BACKUP_NO_SHARE',
  BACKUP_COMPLETE: 'BACKUP_COMPLETE',
}

export type Passphrase = {
  phrase: string,
  salt: string,
  hash: string,
}

export type PrepareBackupStatus =
  | typeof PREPARE_BACK_IDLE
  | typeof PREPARE_BACKUP_LOADING
  | typeof PREPARE_BACKUP_SUCCESS
  | typeof PREPARE_BACKUP_FAILURE

export type BackupStore = {
  passphrase: Passphrase,
  backupWalletPath: string,
  showBanner: boolean,
  lastSuccessfulBackup: string,
  // TODO: fix flow type
  error: any,
  status: string,
  prepareBackupStatus: PrepareBackupStatus,
}

export type StoreError = { error: ?CustomError }
export type BackupStoreStatus = { status: $Keys<typeof BACKUP_STORE_STATUS> }

export const ERROR_EXPORT_BACKUP = {
  code: 'WB-001',
  message: 'Error while exporting backup file',
}

export const ERROR_GENERATE_BACKUP_FILE = {
  code: 'WB-002',
  message: 'Error while generating backup file',
}

export const ERROR_GENERATE_RECOVERY_PHRASE = {
  code: 'WB-003',
  message: 'Error while generating recovery phrase',
}

export const ERROR_HYDRATING_BACKUP = {
  code: 'WB-004',
  message: 'Error hydrating backup',
}

export const START_BACKUP = 'START_BACKUP'
export const GENERATE_BACKUP_FILE_LOADING = 'GENERATE_BACKUP_FILE_LOADING'
export const GENERATE_BACKUP_FILE_SUCCESS = 'GENERATE_BACKUP_FILE_SUCCESS'
export const GENERATE_BACKUP_FILE_FAILURE = 'GENERATE_BACKUP_FILE_FAILURE'
export const BACKUP_WALLET_FAIL = 'BACKUP_WALLET_FAIL'
export const GENERATE_RECOVERY_PHRASE_LOADING =
  'GENERATE_RECOVERY_PHRASE_LOADING'
export const GENERATE_RECOVERY_PHRASE_SUCCESS =
  'GENERATE_RECOVERY_PHRASE_SUCCESS'
export const GENERATE_RECOVERY_PHRASE_FAILURE =
  'GENERATE_RECOVERY_PHRASE_FAILURE'
export const EXPORT_BACKUP_LOADING = 'EXPORT_BACKUP_LOADING'
export const EXPORT_BACKUP_SUCCESS = 'EXPORT_BACKUP_SUCCESS'
export const EXPORT_BACKUP_FAILURE = 'EXPORT_BACKUP_FAILURE'
export const EXPORT_BACKUP_NO_SHARE = 'EXPORT_BACKUP_NO_SHARE'
export const BACKUP_COMPLETE = 'BACKUP_COMPLETE'
export const HYDRATE_BACKUP = 'HYDRATE_BACKUP'
export const HYDRATE_BACKUP_FAILURE = 'HYDRATE_BACKUP_FAILURE'
export const PROMPT_WALLET_BACKUP_BANNER = 'PROMPT_WALLET_BACKUP_BANNER'
export const WALLET_FILE_NAME = 'ConnectMe'

export type BackupStartAction = {
  type: typeof START_BACKUP,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
  error: CustomError,
}

export type GenerateBackupFileLoadingAction = {
  type: typeof GENERATE_BACKUP_FILE_LOADING,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
}

export type GenerateBackupFileSuccessAction = {
  type: typeof GENERATE_BACKUP_FILE_SUCCESS,
  backupWalletPath: string,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
}

export type GenerateBackupFileFailureAction = {
  type: typeof GENERATE_BACKUP_FILE_FAILURE,
  error: CustomError,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
}

export type BackupWalletFailAction = {
  type: typeof BACKUP_WALLET_FAIL,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
  error: CustomError,
}

export type GenerateRecoveryPhraseLoadingAction = {
  type: typeof GENERATE_RECOVERY_PHRASE_LOADING,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
}

export type GenerateRecoveryPhraseSuccessAction = {
  type: typeof GENERATE_RECOVERY_PHRASE_SUCCESS,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
  passphrase: Passphrase,
  error: CustomError,
}

export type GenerateRecoveryPhraseFailureAction = {
  type: typeof GENERATE_RECOVERY_PHRASE_FAILURE,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
  error: CustomError,
}

export type ExportBackupLoadingAction = {
  type: typeof EXPORT_BACKUP_LOADING,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
  backupWalletPath: string,
}

export type PrepareBackupLoadingAction = {
  type: typeof PREPARE_BACKUP_LOADING,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
}

export type ExportBackupNoShareAction = {
  type: typeof EXPORT_BACKUP_NO_SHARE,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
}

export type ExportBackupSuccessAction = {
  type: typeof EXPORT_BACKUP_SUCCESS,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
  lastSuccessfulBackup: string,
}

export type ExportBackupFailureAction = {
  type: typeof EXPORT_BACKUP_FAILURE,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
  error: CustomError,
}

export type BackupCompleteAction = {
  type: typeof BACKUP_COMPLETE,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
}

export type HydrateBackupAction = {
  type: typeof HYDRATE_BACKUP,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
  lastSuccessfulBackup: string,
}

export type HydrateBackupFailAction = {
  type: typeof HYDRATE_BACKUP_FAILURE,
  error: CustomError,
}

export type PromptBackupBannerAction = {
  type: typeof PROMPT_WALLET_BACKUP_BANNER,
  showBanner: boolean,
}

export type PrepareBackupSuccessAction = {
  type: typeof PREPARE_BACKUP_SUCCESS,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
}

export type PrepareBackupFailureAction = {
  type: typeof PREPARE_BACKUP_FAILURE,
  status: $Keys<typeof BACKUP_STORE_STATUS>,
}

export type BackupStoreAction =
  | BackupStartAction
  | GenerateBackupFileLoadingAction
  | GenerateBackupFileSuccessAction
  | BackupWalletFailAction
  | GenerateRecoveryPhraseLoadingAction
  | GenerateRecoveryPhraseSuccessAction
  | GenerateBackupFileFailureAction
  | PrepareBackupLoadingAction
  | ExportBackupLoadingAction
  | ExportBackupSuccessAction
  | ExportBackupFailureAction
  | ExportBackupNoShareAction
  | BackupCompleteAction
  | HydrateBackupAction
  | HydrateBackupFailAction
  | PrepareBackupSuccessAction
  | PrepareBackupFailureAction
