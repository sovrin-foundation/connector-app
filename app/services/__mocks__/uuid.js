// @flow
import { defaultUUID } from '../../../__mocks__/static-data'

export function uuid() {
  return defaultUUID
}
