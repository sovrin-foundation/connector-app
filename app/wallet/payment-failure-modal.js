// @flow
import React, { PureComponent } from 'react'
import { StyleSheet, Platform } from 'react-native'
import {
  CustomView,
  Icon,
  CustomText,
  CustomButton,
  Container,
} from '../components'
import {
  OFFSET_1X,
  OFFSET_2X,
  OFFSET_3X,
  OFFSET_6X,
  OFFSET_7X,
  color,
  HAIRLINE_WIDTH,
  whiteSmoke,
} from '../common/styles/constant'
import Modal from 'react-native-modal'

type PaymentFailureModalProps = {
  testID: string,
  onClose: () => void,
  onRetry: (event: any) => void,
  connectionName: string,
}

export default class PaymentFailureModal extends PureComponent<
  PaymentFailureModalProps,
  void
> {
  render() {
    const { testID, onClose } = this.props
    return (
      <CustomView fifth shadow style={[styles.container]}>
        <CustomView spaceBetween style={[styles.innerContainer]}>
          <Icon
            iconStyle={[{ margin: 10 }]}
            src={require('../images/alertInfo.png')}
            extraLarge
            center
            resizeMode="contain"
            testID={`${testID}-modal-header-icon`}
          />
          <CustomText
            h5
            center
            tertiary
            bg="tertiary"
            transparentBg
            style={[styles.message]}
            demiBold
            testID={`${testID}-modal-title`}
          >
            {'Payment Failure'}
          </CustomText>
          <CustomText
            h5
            center
            tertiary
            bg="tertiary"
            transparentBg
            style={[styles.message]}
            demiBold
            testID={`${testID}-modal-content`}
          >
            {`Something went wrong trying to pay ${
              this.props.connectionName
            }. Please try again.`}
          </CustomText>
        </CustomView>
        <CustomView row>
          <Container>
            <CustomButton
              fifth
              onPress={onClose}
              title={'Cancel'}
              testID={`${testID}-modal-cancel`}
              textStyle={{ fontWeight: 'bold' }}
            />
          </Container>
          <Container>
            <CustomButton
              fifth
              onPress={this.props.onRetry}
              title={'Retry'}
              testID={`${testID}-modal-retry`}
              textStyle={{ fontWeight: 'bold' }}
            />
          </Container>
        </CustomView>
      </CustomView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    marginHorizontal: OFFSET_3X,
  },
  innerContainer: {
    ...Platform.select({
      ios: {
        borderBottomColor: color.bg.fifth.font.tertiary,
        borderBottomWidth: HAIRLINE_WIDTH,
      },
      android: {
        borderBottomColor: color.bg.fifth.font.secondary,
        borderBottomWidth: 1,
      },
    }),
    padding: OFFSET_2X,
  },
  message: {
    marginBottom: OFFSET_1X / 2,
  },
})
