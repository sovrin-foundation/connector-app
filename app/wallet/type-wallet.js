// @flow
import type { ReactNavigation } from '../common/type-common'
import type { CustomError } from '../common/type-common'
import type { IsValid } from '../components/input-control/type-input-control'
import type { CredentialOfferModalStatus } from '../claim-offer/type-claim-offer'

export type BackupWalletProps = {
  backup: BackupInfo,
  render: (backup: string, backupWallet: () => void) => void,
  walletBackup: () => void,
} & ReactNavigation

export type BackupInfo = {
  latest: ?string,
  error: ?CustomError,
  backupPath: ?string,
} & StoreStatus

export const BACKUP_WALLET = 'BACKUP_WALLET'

export const BACKUP_WALLET_FAIL = 'BACKUP_WALLET_FAIL'

export const BACKUP_WALLET_SUCCESS = 'BACKUP_WALLET_SUCCESS'

export const GET_WALLET_ENCRYPTION_KEY = 'GET_WALLET_ENCRYPTION_KEY'

export const BACKUP_WALLET_PATH = 'BACKUP_WALLET_PATH'

export const SHARE_WALLET_BACKUP = 'SHARE_WALLET_BACKUP'

export const ERROR_BACKUP_WALLET = {
  code: 'WB-001',
  message: 'Error while backing up wallet',
}

export const ERROR_BACKUP_WALLET_SHARE = {
  code: 'WB-002',
  message: 'Error while sharing zipped backup',
}

export type WalletTabSendDetailsProps = {
  tokenAmount: string,
  tokenSentStatus: $Keys<typeof STORE_STATUS>,
  sendTokens: (
    tokenAmount: string,
    recipientAddress: string
  ) => SendTokensAction,
} & ReactNavigation

export type BackupWalletAction = {
  type: typeof BACKUP_WALLET,
  data: BackupInfo,
}

export type ShareBackupAction = {
  type: typeof SHARE_WALLET_BACKUP,
  data: BackupInfo,
}

export type WalletBalanceProps = {
  walletBalance: string,
  refreshWalletBalance: () => {},
  render: Function,
}

export type WalletSendAmountState = {
  text: string,
}

export type WalletSendAmountProps = {
  screenProps: ReactNavigation,
  selectTokenAmount: (tokenAmount: string) => {},
  walletBalance: string,
  paymentStatus: $Keys<typeof STORE_STATUS>,
} & ReactNavigation

export type WalletProps = {} & ReactNavigation

export type WalletPayload = {
  walletPath: string,
  encryptionKey: string,
}

export type WalletTabReceiveProps = {
  walletAddresses: Array<string>,
  refreshWalletAddresses: () => {},
  promptBackupBanner: (showBackup: boolean) => {},
  addressStatus: $Keys<typeof STORE_STATUS>,
  currentScreen: string,
}

export type WalletTabReceiveState = {
  copyButtonText: string,
}

export type WalletSendPaymentData = {
  paymentTo: string,
  paymentFor?: string,
}

export type WalletTabSendDetailsState = {
  showPaymentAddress: boolean,
  isPaymentAddressValid: IsValid,
  credentialOfferModalStatus: CredentialOfferModalStatus,
}

export type WalletHistoryProps = {
  walletHistory: WalletHistory,
  refreshWalletHistory: () => {},
}

export type WalletTabsProps = {} & ReactNavigation

export type WalletHistoryEvent = {
  id: string,
  senderName?: string,
  senderAddress: string,
  action: string,
  tokenAmount: string,
  timeStamp: string,
}

export type WalletBalance = {
  data: string,
} & StoreStatus &
  StoreError

export type WalletAddresses = {
  data: Array<string>,
} & StoreStatus &
  StoreError

export type WalletHistory = {
  transactions: WalletHistoryEvent[],
} & StoreStatus &
  StoreError

export type Payment = {
  tokenAmount: string,
} & StoreStatus &
  StoreError

export const ERROR_LOADING_WALLET = {
  code: 'W001',
  message: 'Error while loading wallet data',
}

export const ERROR_REFRESHING_WALLET_BALANCE = {
  code: 'W002',
  message: 'Error while refreshing wallet balance',
}

export const ERROR_REFRESHING_WALLET_ADDRESSES = {
  code: 'W003',
  message: 'Error while refreshing wallet addresses',
}

export const ERROR_REFRESHING_WALLET_HISTORY = {
  code: 'W004',
  message: 'Error while refreshing wallet history',
}

export const ERROR_LOADING_WALLET_BALANCE = {
  code: 'W005',
  message: 'Error while loading wallet balance',
}

export const ERROR_LOADING_WALLET_ADDRESSES = {
  code: 'W006',
  message: 'Error while loading wallet addresses',
}

export const ERROR_LOADING_WALLET_HISTORY = {
  code: 'W007',
  message: 'Error while loading wallet history',
}

export const ERROR_SENDING_TOKENS = {
  code: 'W008',
  message: 'Error sending tokens',
}

export const ERROR_SENDING_TOKENS_WITH_FEES = {
  code: 'W009',
  message:
    'Error sending tokens while transfer fees is greater than or equal to transfer amount',
}

export const STORE_STATUS = {
  IDLE: 'IDLE',
  IN_PROGRESS: 'IN_PROGRESS',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
}

export type StoreError = { error: ?CustomError }
export type StoreStatus = { status: $Keys<typeof STORE_STATUS> }

export type WalletStore = {
  walletBalance: WalletBalance,
  walletAddresses: WalletAddresses,
  walletHistory: WalletHistory,
  backup: BackupInfo,
  payment: Payment,
}

export type HydrateWalletBalanceFailAction = {
  type: typeof HYDRATE_WALLET_BALANCE_FAIL,
  error: CustomError,
  status: StoreStatus,
}

export type HydrateWalletAddressesFailAction = {
  type: typeof HYDRATE_WALLET_ADDRESSES_FAIL,
  error: CustomError,
  status: StoreStatus,
}

export type HydrateWalletHistoryFailAction = {
  type: typeof HYDRATE_WALLET_HISTORY_FAIL,
  error: CustomError,
  status: StoreStatus,
}

export type SendTokensAction = {
  type: typeof SEND_TOKENS,
  tokenAmount: string,
  recipientWalletAddress: string,
}

export type SelectTokensAction = {
  type: typeof SEND_TOKENS,
  tokenAmount: string,
}

export const HYDRATE_WALLET_BALANCE_FAIL = 'HYDRATE_WALLET_BALANCE_FAIL'
export const HYDRATE_WALLET_ADDRESSES_FAIL = 'HYDRATE_WALLET_ADDRESSES_FAIL'
export const HYDRATE_WALLET_HISTORY_FAIL = 'HYDRATE_WALLET_HISTORY_FAIL'
export const REFRESH_WALLET_BALANCE_FAIL = 'REFRESH_WALLET_BALANCE_FAIL'
export const REFRESH_WALLET_ADDRESSES_FAIL = 'REFRESH_WALLET_ADDRESSES_FAIL'
export const REFRESH_WALLET_HISTORY_FAIL = 'REFRESH_WALLET_HISTORY_FAIL'
export const HYDRATE_WALLET_BALANCE = 'HYDRATE_WALLET_BALANCE'
export const HYDRATE_WALLET_ADDRESSES = 'HYDRATE_WALLET_ADDRESSES'
export const HYDRATE_WALLET_HISTORY = 'HYDRATE_WALLET_HISTORY'
export const REFRESH_WALLET = 'REFRESH_WALLET'
export const REFRESH_WALLET_ADDRESSES = 'REFRESH_WALLET_ADDRESSES'
export const REFRESH_WALLET_HISTORY = 'REFRESH_WALLET_HISTORY'
export const REFRESH_WALLET_BALANCE = 'REFRESH_WALLET_BALANCE'
export const WALLET_BALANCE_REFRESHED = 'WALLET_BALANCE_REFRESHED'
export const WALLET_ADDRESSES_REFRESHED = 'WALLET_ADDRESSES_REFRESHED'
export const WALLET_HISTORY_REFRESHED = 'WALLET_HISTORY_REFRESHED'
export const SEND_TOKENS = 'SEND_TOKENS'
export const SEND_TOKENS_FAIL = 'SEND_TOKENS_FAIL'
export const TOKEN_SENT_SUCCESS = 'TOKEN_SENT_SUCCESS'
export const SELECT_TOKEN_AMOUNT = 'SELECT_TOKEN_AMOUNT'
export const PROMPT_WALLET_BACKUP_BANNER = 'PROMPT_WALLET_BACKUP_BANNER'
export const WALLET_ADDRESSES_FETCH_START = 'WALLET_ADDRESSES_FETCH_START'

export type PromptBackupBannerAction = {
  type: typeof PROMPT_WALLET_BACKUP_BANNER,
  showBanner: boolean,
}

export type HydrateWalletBalanceData = string

export type HydrateWalletAddressesData = Array<string>

export type HydrateWalletHistoryTransactions = WalletHistoryEvent[]

export type RefreshWalletHistoryAction = {
  type: typeof REFRESH_WALLET_HISTORY,
}

export type WalletBalanceRefreshedAction = {
  type: typeof WALLET_BALANCE_REFRESHED,
  walletBalance: WalletBalance,
}

export type WalletAddressesRefreshedAction = {
  type: typeof WALLET_ADDRESSES_REFRESHED,
  walletAddresses: WalletAddresses,
}

export type WalletHistoryRefreshedAction = {
  type: typeof WALLET_HISTORY_REFRESHED,
  walletHistory: WalletHistory,
}

export type RefreshWalletAddressesAction = {
  type: typeof REFRESH_WALLET_ADDRESSES,
}

export type RefreshWalletBalanceAction = {
  type: typeof REFRESH_WALLET_BALANCE,
  walletBalance: WalletBalance,
}

export type HydrateWalletBalanceAction = {
  type: typeof HYDRATE_WALLET_BALANCE,
  walletBalance: WalletBalance,
}

export type HydrateWalletAddressesAction = {
  type: typeof HYDRATE_WALLET_ADDRESSES,
  walletAddresses: WalletAddresses,
}

export type HydrateWalletHistoryAction = {
  type: typeof HYDRATE_WALLET_HISTORY,
  walletHistory: WalletHistory,
}

export type TokenSentSuccessAction = {
  type: typeof TOKEN_SENT_SUCCESS,
  payment: Payment,
}

export type WalletStoreAction =
  | HydrateWalletBalanceFailAction
  | HydrateWalletAddressesFailAction
  | HydrateWalletHistoryFailAction
  | RefreshWalletBalanceAction
  | RefreshWalletAddressesAction
  | RefreshWalletHistoryAction
  | WalletHistoryRefreshedAction
  | WalletBalanceRefreshedAction
  | WalletAddressesRefreshedAction
  | HydrateWalletBalanceAction
  | HydrateWalletAddressesAction
  | HydrateWalletHistoryAction
  | BackupWalletAction
  | SendTokensAction
  | TokenSentSuccessAction
  | SelectTokensAction
