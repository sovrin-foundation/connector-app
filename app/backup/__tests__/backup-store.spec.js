// @flow

import backupReducer, {
  generateBackupFile,
  generateRecoveryPhrase,
  generateRecoveryPhraseSuccess,
  generateBackupFileSuccess,
  exportBackup,
  exportBackupSuccess,
  promptBackupBanner,
  prepareBackup,
  prepareBackupSuccess,
  prepareBackupFail,
} from '../backup-store'
import { BACKUP_STORE_STATUS } from '../type-backup'
import { getStore } from '../../../__mocks__/static-data'
import { backupWalletFail } from '../../wallet/wallet-store'
import { ERROR_BACKUP_WALLET } from '../../wallet/type-wallet'
describe('store: backup-store: ', () => {
  let initialState
  beforeEach(() => {
    initialState = {
      passphrase: { phrase: '', salt: 's', hash: 'h' },
      status: BACKUP_STORE_STATUS.IDLE,
      error: null,
      showBanner: false,
      lastSuccessfulBackup: '',
      backupWalletPath: '',
      prepareBackupStatus: BACKUP_STORE_STATUS.PREPARE_BACK_IDLE,
    }
  })
  const {
    backupWalletPath,
    passphrase,
    lastSuccessfulBackup,
  } = getStore().getState().backup
  const generateRecoveryPhraseState = backupReducer(
    initialState,
    generateRecoveryPhrase()
  )
  const generateRecoveryPhraseSuccessState = backupReducer(
    generateRecoveryPhraseState,
    generateRecoveryPhraseSuccess(passphrase)
  )
  const generateBackupFileState = backupReducer(
    generateRecoveryPhraseState,
    generateBackupFile()
  )
  const generateBackupFileSuccessState = backupReducer(
    generateBackupFileState,
    generateBackupFileSuccess(backupWalletPath)
  )
  const exportBackupState = backupReducer(
    generateBackupFileSuccessState,
    exportBackup()
  )

  it('action: GENERATE_RECOVERY_PHRASE', () => {
    expect(
      backupReducer(initialState, generateRecoveryPhrase())
    ).toMatchSnapshot()
  })

  it('action: GENERATE_RECOVERY_PHRASE_SUCCESS', () => {
    expect(
      backupReducer(
        generateRecoveryPhraseState,
        generateRecoveryPhraseSuccess(passphrase)
      )
    ).toMatchSnapshot()
  })

  it('action: GENERATE_BACKUP_FILE', () => {
    expect(
      backupReducer(generateRecoveryPhraseSuccessState, generateBackupFile())
    ).toMatchSnapshot()
  })

  it('action: GENERATE_BACKUP_FILE_SUCCESS', () => {
    expect(
      backupReducer(
        generateBackupFileState,
        generateBackupFileSuccess(backupWalletPath)
      )
    ).toMatchSnapshot()
  })

  it('action: EXPORT_BACKUP', () => {
    expect(
      backupReducer(generateBackupFileSuccessState, exportBackup())
    ).toMatchSnapshot()
  })

  it('action: EXPORT_BACKUP_SUCCESS', () => {
    expect(
      backupReducer(
        exportBackupState,
        exportBackupSuccess(lastSuccessfulBackup)
      )
    ).toMatchSnapshot()
  })

  // TODO: fix flow type
  // it('action: PROMPT_WALLET_BACKUP_BANNER', () => {
  //   expect(
  //     backupReducer(initialState, promptBackupBanner(true))
  //   ).toMatchSnapshot()
  // })

  it('action: BACKUP_WALLET_FAIL', () => {
    expect(
      backupReducer(initialState, backupWalletFail(ERROR_BACKUP_WALLET))
    ).toMatchSnapshot()
  })

  it('action: PREPARE_BACKUP_LOADING', () => {
    expect(backupReducer(initialState, prepareBackup())).toMatchSnapshot()
  })

  it('action: PREPARE_BACKUP_SUCCESS', () => {
    expect(
      backupReducer(initialState, prepareBackupSuccess())
    ).toMatchSnapshot()
  })

  it('action: PREPARE_BACKUP_FAILURE', () => {
    const testError = {
      code: '000',
      message: 'test message',
    }
    expect(
      backupReducer(initialState, prepareBackupFail(testError))
    ).toMatchSnapshot()
  })

  // TODO:KS Start adding tests for all sagas inside backup store
})
