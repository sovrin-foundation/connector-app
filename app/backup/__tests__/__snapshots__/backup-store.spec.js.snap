// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store: backup-store:  action: BACKUP_WALLET_FAIL 1`] = `
Object {
  "backupWalletPath": "",
  "error": Object {
    "code": "WB-001",
    "message": "Error while backing up wallet",
  },
  "lastSuccessfulBackup": "",
  "passphrase": Object {
    "hash": "h",
    "phrase": "",
    "salt": "s",
  },
  "prepareBackupStatus": "PREPARE_BACKUP_IDLE",
  "showBanner": false,
  "status": "IDLE",
}
`;

exports[`store: backup-store:  action: EXPORT_BACKUP 1`] = `
Object {
  "backupWalletPath": "/Users/CoreSimulator/Devices/5F89AA14-B694-465F-82C3-838552AB4B85/data/Containers/Data/Application/4536A441-4B70-4A30-8DCC-5274E6781535/Documents/backup.zip",
  "error": null,
  "lastSuccessfulBackup": "",
  "passphrase": Object {
    "hash": "h",
    "phrase": "",
    "salt": "s",
  },
  "prepareBackupStatus": "PREPARE_BACKUP_IDLE",
  "showBanner": false,
  "status": "EXPORT_BACKUP_LOADING",
}
`;

exports[`store: backup-store:  action: EXPORT_BACKUP_SUCCESS 1`] = `
Object {
  "backupWalletPath": "/Users/CoreSimulator/Devices/5F89AA14-B694-465F-82C3-838552AB4B85/data/Containers/Data/Application/4536A441-4B70-4A30-8DCC-5274E6781535/Documents/backup.zip",
  "error": null,
  "lastSuccessfulBackup": "2018-06-27T18:27:35+05:30",
  "passphrase": Object {
    "hash": "h",
    "phrase": "",
    "salt": "s",
  },
  "prepareBackupStatus": "PREPARE_BACKUP_IDLE",
  "showBanner": false,
  "status": "BACKUP_COMPLETE",
}
`;

exports[`store: backup-store:  action: GENERATE_BACKUP_FILE 1`] = `
Object {
  "backupWalletPath": "",
  "error": null,
  "lastSuccessfulBackup": "",
  "passphrase": Object {
    "hash": "b7a563bcbbeb892b",
    "phrase": "gander troubling deodorize justify twitter darkish confront politely",
    "salt": "s",
  },
  "prepareBackupStatus": "PREPARE_BACKUP_IDLE",
  "showBanner": false,
  "status": "GENERATE_BACKUP_FILE_LOADING",
}
`;

exports[`store: backup-store:  action: GENERATE_BACKUP_FILE_SUCCESS 1`] = `
Object {
  "backupWalletPath": "/Users/CoreSimulator/Devices/5F89AA14-B694-465F-82C3-838552AB4B85/data/Containers/Data/Application/4536A441-4B70-4A30-8DCC-5274E6781535/Documents/backup.zip",
  "error": null,
  "lastSuccessfulBackup": "",
  "passphrase": Object {
    "hash": "h",
    "phrase": "",
    "salt": "s",
  },
  "prepareBackupStatus": "PREPARE_BACKUP_IDLE",
  "showBanner": false,
  "status": "GENERATE_BACKUP_FILE_SUCCESS",
}
`;

exports[`store: backup-store:  action: GENERATE_RECOVERY_PHRASE 1`] = `
Object {
  "backupWalletPath": "",
  "error": null,
  "lastSuccessfulBackup": "",
  "passphrase": Object {
    "hash": "h",
    "phrase": "",
    "salt": "s",
  },
  "prepareBackupStatus": "PREPARE_BACKUP_IDLE",
  "showBanner": false,
  "status": "GENERATE_PHRASE_LOADING",
}
`;

exports[`store: backup-store:  action: GENERATE_RECOVERY_PHRASE_SUCCESS 1`] = `
Object {
  "backupWalletPath": "",
  "error": undefined,
  "lastSuccessfulBackup": "",
  "passphrase": Object {
    "hash": "b7a563bcbbeb892b",
    "phrase": "gander troubling deodorize justify twitter darkish confront politely",
    "salt": "s",
  },
  "prepareBackupStatus": "PREPARE_BACKUP_IDLE",
  "showBanner": false,
  "status": "GENERATE_PHRASE_SUCCESS",
}
`;

exports[`store: backup-store:  action: PREPARE_BACKUP_FAILURE 1`] = `
Object {
  "backupWalletPath": "",
  "error": null,
  "lastSuccessfulBackup": "",
  "passphrase": Object {
    "hash": "h",
    "phrase": "",
    "salt": "s",
  },
  "prepareBackupStatus": "PREPARE_BACKUP_FAILURE",
  "showBanner": false,
  "status": "IDLE",
}
`;

exports[`store: backup-store:  action: PREPARE_BACKUP_LOADING 1`] = `
Object {
  "backupWalletPath": "",
  "error": null,
  "lastSuccessfulBackup": "",
  "passphrase": Object {
    "hash": "h",
    "phrase": "",
    "salt": "s",
  },
  "prepareBackupStatus": "PREPARE_BACKUP_LOADING",
  "showBanner": false,
  "status": "IDLE",
}
`;

exports[`store: backup-store:  action: PREPARE_BACKUP_SUCCESS 1`] = `
Object {
  "backupWalletPath": "",
  "error": null,
  "lastSuccessfulBackup": "",
  "passphrase": Object {
    "hash": "h",
    "phrase": "",
    "salt": "s",
  },
  "prepareBackupStatus": "PREPARE_BACKUP_SUCCESS",
  "showBanner": false,
  "status": "IDLE",
}
`;
